File: 0589-n-ary-tree-preorder-traversal/0589-n-ary-tree-preorder-traversal.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<Integer> preorder(Node root) {
      List<Integer> ans = new LinkedList<>();
      if(root == null) return ans;
      
      
      Stack<Node> S = new Stack<Node>();
      Node t;
      S.push(root);
      while(!S.isEmpty()) {
         t = S.pop();
        ans.add(t.val);
        for(int i = t.children.size() - 1; i >= 0; --i) {
          S.push(t.children.get(i));
        }
      }
      
      return ans;
    }
}

File: 0589-n-ary-tree-preorder-traversal/0589-n-ary-tree-preorder-traversal.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

// Recursive
// class Solution {
// public:
//     vector<int> ans;
//     void traverse(Node* root) {
//       if (root) {
//         ans.push_back(root->val);
//         for (Node* child: root->children) // for in
//           traverse(child);
//       }
//     }
//     vector<int> preorder(Node* root) {
//         traverse(root);
//         return ans;
//     }
// };

// Iterative
class Solution {
public:
    vector<int> ans;
    stack<Node*> stack;
    vector<int> preorder(Node* root) {
        if (!root) return {};
        
        Node* t;
        stack.push(root);
        while(!stack.empty()) {
          t = stack.top();
          stack.pop();
          
          ans.push_back(t->val); // add to ans
          
          // push stack from right to left, so they leave from left to right
          for(int i = t->children.size() - 1; i >= 0; --i) 
            stack.push(t->children[i]);
        }
        
        return ans;
    }
};